### PR TITLE
Fix getCentroid for fragmented features and rename to getRenderedCentroid

### PR DESCRIPTION
**Pull request by VictorVelarde**
### Related to https://github.com/CartoDB/carto-vl/issues/1215

*Note*: starting from a Mamata's branch (nothing remarkable for the current issue). 

See the discussion below ⬇️ starting [here](https://github.com/CartoDB/carto-vl/pull/1217#issuecomment-451899937)

### CODE REVIEW

**Comment by VictorVelarde** [comment]
##  1. The problem
The getCentroid is attached to Features coming from:
- `viewportFeatures`
- `Interactivity` (to be analyzed)

After reviewing the calculations for a while, a situation has finally arisen: the `getCentroid` operation currently works at the **dataframe level**, but there is no mechanism yet to consider the different 'pieces' (that could lie in several dataframes) and how the centroid should be calculated for the **whole set of geometries**.

### a) viewportFeatures
This is the current code in `ViewportAggCalculator.js` which runs the viewport aggregations:
```js
const featureAlreadyAccumulated = processedFeaturesIDs.has(featureId);
if (featureAlreadyAccumulated) {
    continue;
}
```
So if the feature in the dataframe (a piece of it) has already been included in the list of features to be processed later by the viewport expressions, new occurrences are ignored. That's fine for alphanumeric properties (and performance), but causes that, when using viewportFeatures, the *first piece of feature (and its centroid) wins*, so that's the feature & centroid it is recovered later on. The current calculation for polygons looks fine but just when considering it for a single dataframe, which is not so good.

Commenting out the previous lines it is not a solution, but it displays the situation more clearly (displaying the centroid of all the 'pieces'):
several_centroids

That's why:
- This doesn't happen with GeoJSON data sources (1 dataframe), like the ones used in the debug examples.
- This doesn't manifest so clearly when using a lower zoom (when the features lie inside just one dataframe)
- This doesn't affect point geometries, which always fit inside just 1 dataframe.

### b) Interactivity

This gif shows the centroid obtained from a featureClick event, with Interactivity
centroid-interactivity

The problem is similar, in the sense that it just returns the centroid for a "portion of the feature" (the one under the click coordinates).

In this case query method is `getFeaturesAtPosition` at `RenderLayer`.


## 2. Approach
- Recalculate the centroids considering the whole geometry, for lines and polygons:
    - [x] viewportFeatures 
    - [ ] Interactivity
- Rename `getCentroid` to something like `getClientCentroid`, to better convey the idea of 'client-side calculation' (just considering visible tiles, not the whole feature in the server).

**Comment by VictorVelarde** [comment]
#### Performance for viewportFeatures
- Points: are is not affected
- Lines / Polygons:
    - eg */hurricane-harvey.html* (about 2500 polygons animation). Adding previously a `@v_features: viewportFeatures($e_totpop)` to the viz, there is no remarkable drop in FPS.
    - eg. */debug/layer-updated.html* (about 29000 polygons anim). Adding previously `@v_features: viewportFeatures($beginning)` to the animation, drops from 11-12FPS to 10-11 FPS aprox.

So it's clear it needs more calculations, and performance must be a bit slower, but it doesn't look like a blocker.

**Comment by VictorVelarde** [comment]
### Fix results

#### Step 1
Features with multiple feature pieces
image

#### Step 2
Multiple pieces, ponderated by area
image

File: src/renderer/RenderLayer.js [hunk 1]
```diff
@@ -60,17 +63,90 @@ export default class RenderLayer {
...
+        }
+        return features;
+    }
+
+    /**
+     * Get all the pieces from rawFeatures.
+     * Return is a nested Array; each entry includes an array with (potentially) several feature pieces
+     */
+    _getPartialFeaturesFromSingle (rawFeatures) {
+        const idProperty = this.viz.metadata.idProperty;
```
Comment (rochoa) [inline src/renderer/RenderLayer.js]: Should we use `this.idProperty`? Or is it unsafe?

File: src/interactivity/commonFeature.js [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+/**
+ * Get a 'compound Feature'. It is a Feature created from a set of 'feature' pieces
+ * @param {Array} featurePieces
+ */
+export function getCompoundFeature (featurePieces) {
+    const exemplar = featurePieces[0];
+    if (featurePieces.length === 1) return exemplar;
+
+    const geomProperties = _getGeomPropertiesFromPieces(featurePieces);
+    Object.defineProperty(exemplar, 'getCentroid', {
```
Comment (rochoa) [inline src/interactivity/commonFeature.js]: What's the reason to use a getter here?

Comment (VictorVelarde) [reply src/interactivity/commonFeature.js]: I was trying to gain some performance here, by lazy calculating some stuff just when invoking the method, the same as when creating a Feature from scratch

File: src/renderer/RenderLayer.js (hunk 1 above)
Comment (VictorVelarde) [reply src/renderer/RenderLayer.js]: Good point, let me check

**Pull request merged**